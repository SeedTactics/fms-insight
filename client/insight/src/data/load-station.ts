/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import * as api from "./api";
import { HashMap, Vector } from "prelude-ts";

export interface PalletData {
  pallet: api.IPalletStatus;
  material: api.IInProcessMaterial[];
}

export function buildPallets(
  st: Readonly<api.ICurrentStatus>
): HashMap<string, { pal?: PalletData; queued?: PalletData }> {
  const matByPallet = new Map<string, api.IInProcessMaterial[]>();
  for (const mat of st.material) {
    if (mat.location.type === api.LocType.OnPallet && mat.location.pallet !== undefined) {
      const mats = matByPallet.get(mat.location.pallet) || [];
      mats.push(mat);
      matByPallet.set(mat.location.pallet, mats);
    }
  }

  const m = new Map<string, { pal?: PalletData; queued?: PalletData }>();
  for (const pal of Object.values(st.pallets)) {
    switch (pal.currentPalletLocation.loc) {
      case api.PalletLocationEnum.LoadUnload:
      case api.PalletLocationEnum.Machine: {
        const stat = pal.currentPalletLocation.group + " #" + pal.currentPalletLocation.num.toString();
        m.set(stat, {
          ...(m.get(stat) || {}),
          pal: {
            pallet: pal,
            material: matByPallet.get(pal.pallet) || [],
          },
        });
        break;
      }

      case api.PalletLocationEnum.MachineQueue: {
        const stat2 = pal.currentPalletLocation.group + " #" + pal.currentPalletLocation.num.toString();
        m.set(stat2, {
          ...(m.get(stat2) || {}),
          queued: {
            pallet: pal,
            material: matByPallet.get(pal.pallet) || [],
          },
        });
        break;
      }

      // TODO: buffer and cart
    }
  }

  return HashMap.ofIterable(m);
}

export type MaterialList = ReadonlyArray<Readonly<api.IInProcessMaterial>>;

export interface LoadStationAndQueueData {
  readonly loadNum: number;
  readonly pallet?: Readonly<api.IPalletStatus>;
  readonly face: HashMap<number, MaterialList>;
  readonly stationStatus?: HashMap<string, { pal?: PalletData; queued?: PalletData }>;
  readonly freeLoadingMaterial: MaterialList;
  readonly free?: MaterialList;
  readonly queues: HashMap<string, MaterialList>;
}

export function selectLoadStationAndQueueProps(
  loadNum: number,
  queues: ReadonlyArray<string>,
  displayFree: boolean,
  curSt: Readonly<api.ICurrentStatus>
): LoadStationAndQueueData {
  let pal: Readonly<api.IPalletStatus> | undefined;
  if (loadNum >= 0) {
    for (const p of Object.values(curSt.pallets)) {
      if (
        p.currentPalletLocation.loc === api.PalletLocationEnum.LoadUnload &&
        p.currentPalletLocation.num === loadNum
      ) {
        pal = p;
        break;
      }
    }
  }

  let byFace: HashMap<number, api.IInProcessMaterial[]> = HashMap.empty();
  let palName: string | undefined;
  let freeLoading: ReadonlyArray<Readonly<api.IInProcessMaterial>> = [];
  let stationStatus: HashMap<string, { pal?: PalletData; queued?: PalletData }> | undefined;

  // first free and queued material
  let free: MaterialList | undefined;
  if (displayFree) {
    free = curSt.material.filter(
      (m) => m.action.processAfterLoad && m.action.processAfterLoad > 1 && m.location.type === api.LocType.Free
    );
  }

  const queueNames = HashMap.ofIterable<string, api.IInProcessMaterial[]>(
    queues.map((q) => [q, []] as [string, api.IInProcessMaterial[]])
  );
  const queueMat = Vector.ofIterable(curSt.material)
    .filter((m) => {
      if (m.location.type !== api.LocType.InQueue) {
        return false;
      }
      if (queueNames.containsKey(m.location.currentQueue || "")) {
        return true;
      } else if (pal && m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === pal.pallet) {
        return true; // display even if queue not selected
      }
      return false;
    })
    .groupBy((m) => m.location.currentQueue || "")
    .mapValues((ms) => ms.sortOn((m) => m.location.queuePosition || 0).toArray());

  // load pallet material
  if (pal !== undefined) {
    palName = pal.pallet;

    byFace = Vector.ofIterable(curSt.material)
      .filter((m) => m.location.type === api.LocType.OnPallet && m.location.pallet === palName)
      .groupBy((m) => m.location.face || 0)
      .mapValues((ms) => ms.toArray());

    freeLoading = curSt.material.filter(
      (m) =>
        m.action.type === api.ActionType.Loading &&
        m.action.loadOntoPallet === palName &&
        m.action.processAfterLoad === 1 &&
        m.location.type === api.LocType.Free
    );

    // make sure all face desinations exist
    queueMat.forEach(([_, mats]) =>
      mats.forEach((m) => {
        if (m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === palName) {
          const face = m.action.loadOntoFace;
          if (face !== undefined && !byFace.containsKey(face)) {
            byFace = byFace.put(face, []);
          }
        }
      })
    );
    [...freeLoading, ...(free || [])].forEach((m) => {
      if (m.action.type === api.ActionType.Loading && m.action.loadOntoPallet === palName) {
        const face = m.action.loadOntoFace;
        if (face !== undefined && !byFace.containsKey(face)) {
          byFace = byFace.put(face, []);
        }
      }
    });
  } else {
    stationStatus = buildPallets(curSt);
    if (displayFree) {
      freeLoading = curSt.material.filter(
        (m) => m.action.processAfterLoad === 1 && m.location.type === api.LocType.Free
      );
    }
  }

  return {
    loadNum,
    pallet: pal,
    face: byFace,
    stationStatus: stationStatus,
    freeLoadingMaterial: freeLoading,
    free: free,
    queues: queueNames.mergeWith(queueMat, (m1, m2) => [...m1, ...m2]),
  };
}
