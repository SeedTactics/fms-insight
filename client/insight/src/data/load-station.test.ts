/* Copyright (c) 2018, John Lenz

All rights reserved.

Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:

    * Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.

    * Redistributions in binary form must reproduce the above
      copyright notice, this list of conditions and the following
      disclaimer in the documentation and/or other materials provided
      with the distribution.

    * Neither the name of John Lenz, Black Maple Software, SeedTactics,
      nor the names of other contributors may be used to endorse or
      promote products derived from this software without specific
      prior written permission.

THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import curSt from "../../test/status-mock.json";
import { CurrentStatus } from "../network/api.js";
import { selectLoadStationAndQueueProps } from "./load-station.js";
import { describe, it, expect } from "vitest";
import { toRawJs } from "../../test/to-raw-js.js";

describe("load station status", () => {
  it("load 1 with no queues", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(1, [], status))).toMatchSnapshot("load 1 with no queues");
  });

  it("load 2 with queue", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(2, ["Queue1"], status))).toMatchSnapshot(
      "load 2 with queue"
    );
  });

  it("load 3 with empty pallet", () => {
    const status = CurrentStatus.fromJS(curSt);
    expect(toRawJs(selectLoadStationAndQueueProps(3, ["Queue2"], status))).toMatchSnapshot(
      "load 3 with empty pallet"
    );
  });
});
