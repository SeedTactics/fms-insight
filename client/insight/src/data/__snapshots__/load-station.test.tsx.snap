// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`load station status load 1 with no queues: load 1 with no queues 1`] = `
Object {
  "face": Map {
    "1" => Array [
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": 2,
          "LoadOntoPallet": "1",
          "PathAfterLoad": 1,
          "ProcessAfterLoad": 2,
          "Program": undefined,
          "Type": "Loading",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "aaa-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": undefined,
          "Face": 1,
          "Pallet": "1",
          "QueuePosition": undefined,
          "Type": "OnPallet",
        },
        "MaterialID": 100432,
        "PartName": "aaa",
        "Path": 1,
        "Process": 1,
        "Serial": "NBTGI",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
    ],
    "2" => Array [
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "UnloadToCompletedMaterial",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "aaa-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": undefined,
          "Face": 2,
          "Pallet": "1",
          "QueuePosition": undefined,
          "Type": "OnPallet",
        },
        "MaterialID": 100316,
        "PartName": "aaa",
        "Path": 1,
        "Process": 2,
        "Serial": "MZRGY",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
    ],
  },
  "free": undefined,
  "freeLoadingMaterial": Array [
    Object {
      "Action": Object {
        "ElapsedLoadUnloadTime": undefined,
        "ElapsedMachiningTime": undefined,
        "ExpectedRemainingMachiningTime": undefined,
        "LoadOntoFace": 1,
        "LoadOntoPallet": "1",
        "PathAfterLoad": 1,
        "ProcessAfterLoad": 1,
        "Program": undefined,
        "Type": "Loading",
        "UnloadIntoQueue": undefined,
      },
      "JobUnique": "aaa-offline-2018-01-01",
      "LastCompletedMachiningRouteStopIndex": undefined,
      "Location": Object {
        "CurrentQueue": undefined,
        "Face": undefined,
        "Pallet": undefined,
        "QueuePosition": undefined,
        "Type": "Free",
      },
      "MaterialID": 100663,
      "PartName": "bbb",
      "Path": 1,
      "Process": 1,
      "Serial": "NRWGM",
      "SignaledInspections": Array [],
      "WorkorderId": undefined,
    },
  ],
  "loadNum": 1,
  "pallet": Object {
    "CurrentPalletLocation": Object {
      "group": "L/U",
      "loc": "LoadUnload",
      "num": 1,
    },
    "FixtureOnPallet": "fix1",
    "NewFixture": undefined,
    "NumFaces": 2,
    "OnHold": false,
    "Pallet": "1",
    "PercentMoveCompleted": undefined,
    "TargetLocation": undefined,
  },
  "queues": Map {},
  "stationStatus": undefined,
}
`;

exports[`load station status load 2 with queue: load 2 with queue 1`] = `
Object {
  "face": Map {
    "1" => Array [
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "UnloadToCompletedMaterial",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "xxx-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": undefined,
          "Face": 1,
          "Pallet": "5",
          "QueuePosition": undefined,
          "Type": "OnPallet",
        },
        "MaterialID": 102097,
        "PartName": "xxx",
        "Path": 1,
        "Process": 1,
        "Serial": "MBZG4",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
    ],
  },
  "free": undefined,
  "freeLoadingMaterial": Array [],
  "loadNum": 2,
  "pallet": Object {
    "CurrentPalletLocation": Object {
      "group": "L/U",
      "loc": "LoadUnload",
      "num": 2,
    },
    "FixtureOnPallet": "fix5",
    "NewFixture": undefined,
    "NumFaces": 1,
    "OnHold": false,
    "Pallet": "5",
    "PercentMoveCompleted": undefined,
    "TargetLocation": undefined,
  },
  "queues": Map {
    "Queue1" => Array [
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "Waiting",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "aaa-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": "Queue1",
          "Face": undefined,
          "Pallet": undefined,
          "QueuePosition": 0,
          "Type": "InQueue",
        },
        "MaterialID": 104289,
        "PartName": "aaa",
        "Path": 1,
        "Process": 1,
        "Serial": "MRYHE",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "Waiting",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "xxx-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": "Queue1",
          "Face": undefined,
          "Pallet": undefined,
          "QueuePosition": 1,
          "Type": "InQueue",
        },
        "MaterialID": 104458,
        "PartName": "xxx",
        "Path": 1,
        "Process": 1,
        "Serial": "NBVHA",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": 1,
          "LoadOntoPallet": "5",
          "PathAfterLoad": 1,
          "ProcessAfterLoad": 1,
          "Program": undefined,
          "Type": "Loading",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "zzz-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": "Queue1",
          "Face": undefined,
          "Pallet": undefined,
          "QueuePosition": 2,
          "Type": "InQueue",
        },
        "MaterialID": 102304,
        "PartName": "zzz",
        "Path": 1,
        "Process": 1,
        "Serial": "MZQGQ",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
    ],
  },
  "stationStatus": undefined,
}
`;

exports[`load station status load 3 with empty pallet: load 3 with empty pallet 1`] = `
Object {
  "face": Map {},
  "free": undefined,
  "freeLoadingMaterial": Array [],
  "loadNum": 3,
  "pallet": undefined,
  "queues": Map {
    "Queue2" => Array [
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "Waiting",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "ccc-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": "Queue2",
          "Face": undefined,
          "Pallet": undefined,
          "QueuePosition": 0,
          "Type": "InQueue",
        },
        "MaterialID": 104824,
        "PartName": "ccc",
        "Path": 1,
        "Process": 1,
        "Serial": "OBSGQ",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
      Object {
        "Action": Object {
          "ElapsedLoadUnloadTime": undefined,
          "ElapsedMachiningTime": undefined,
          "ExpectedRemainingMachiningTime": undefined,
          "LoadOntoFace": undefined,
          "LoadOntoPallet": undefined,
          "PathAfterLoad": undefined,
          "ProcessAfterLoad": undefined,
          "Program": undefined,
          "Type": "Waiting",
          "UnloadIntoQueue": undefined,
        },
        "JobUnique": "zzz-offline-2018-01-01",
        "LastCompletedMachiningRouteStopIndex": undefined,
        "Location": Object {
          "CurrentQueue": "Queue2",
          "Face": undefined,
          "Pallet": undefined,
          "QueuePosition": 1,
          "Type": "InQueue",
        },
        "MaterialID": 105125,
        "PartName": "zzz",
        "Path": 1,
        "Process": 1,
        "Serial": "MJSGU",
        "SignaledInspections": Array [],
        "WorkorderId": undefined,
      },
    ],
  },
  "stationStatus": Map {
    "Machine #1" => Object {
      "pal": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": "PT10M",
              "ExpectedRemainingMachiningTime": "PT20M",
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": "prog-ccc-2",
              "Type": "Machining",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "ccc-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 2,
              "Pallet": "4",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 101805,
            "PartName": "ccc",
            "Path": 1,
            "Process": 2,
            "Serial": "OBQGU",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": undefined,
              "Type": "Waiting",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "ccc-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "4",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 101662,
            "PartName": "ccc",
            "Path": 1,
            "Process": 1,
            "Serial": "NRWGI",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "Machine",
            "loc": "Machine",
            "num": 1,
          },
          "FixtureOnPallet": "fix4",
          "NewFixture": undefined,
          "NumFaces": 2,
          "OnHold": false,
          "Pallet": "4",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
    },
    "Machine #2" => Object {
      "pal": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": "PT10M",
              "ExpectedRemainingMachiningTime": "PT50M",
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": "prog-zzz",
              "Type": "Machining",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "zzz-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "6",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 102431,
            "PartName": "zzz",
            "Path": 1,
            "Process": 1,
            "Serial": "NBTGE",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "Machine",
            "loc": "Machine",
            "num": 2,
          },
          "FixtureOnPallet": "fix6",
          "NewFixture": undefined,
          "NumFaces": 1,
          "OnHold": false,
          "Pallet": "6",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
      "queued": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": undefined,
              "Type": "Waiting",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "yyy-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "7",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 102715,
            "PartName": "yyy",
            "Path": 1,
            "Process": 1,
            "Serial": "NZRGU",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": undefined,
              "Type": "Waiting",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "yyy-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "7",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 103096,
            "PartName": "yyy",
            "Path": 1,
            "Process": 1,
            "Serial": "MBZGY",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "Machine",
            "loc": "MachineQueue",
            "num": 2,
          },
          "FixtureOnPallet": "fix7",
          "NewFixture": undefined,
          "NumFaces": 1,
          "OnHold": false,
          "Pallet": "7",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
    },
    "Machine #3" => Object {
      "pal": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": "PT10M",
              "ExpectedRemainingMachiningTime": "PT45M",
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": "prog-xxx",
              "Type": "Machining",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "xxx-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "8",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 103335,
            "PartName": "xxx",
            "Path": 1,
            "Process": 1,
            "Serial": "MZTGU",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "Machine",
            "loc": "Machine",
            "num": 3,
          },
          "FixtureOnPallet": "fix8",
          "NewFixture": undefined,
          "NumFaces": 1,
          "OnHold": false,
          "Pallet": "8",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
    },
    "L/U #1" => Object {
      "pal": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": undefined,
              "Type": "UnloadToCompletedMaterial",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "aaa-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 2,
              "Pallet": "1",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 100316,
            "PartName": "aaa",
            "Path": 1,
            "Process": 2,
            "Serial": "MZRGY",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": 2,
              "LoadOntoPallet": "1",
              "PathAfterLoad": 1,
              "ProcessAfterLoad": 2,
              "Program": undefined,
              "Type": "Loading",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "aaa-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "1",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 100432,
            "PartName": "aaa",
            "Path": 1,
            "Process": 1,
            "Serial": "NBTGI",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "L/U",
            "loc": "LoadUnload",
            "num": 1,
          },
          "FixtureOnPallet": "fix1",
          "NewFixture": undefined,
          "NumFaces": 2,
          "OnHold": false,
          "Pallet": "1",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
    },
    "L/U #2" => Object {
      "pal": Object {
        "material": Array [
          Object {
            "Action": Object {
              "ElapsedLoadUnloadTime": undefined,
              "ElapsedMachiningTime": undefined,
              "ExpectedRemainingMachiningTime": undefined,
              "LoadOntoFace": undefined,
              "LoadOntoPallet": undefined,
              "PathAfterLoad": undefined,
              "ProcessAfterLoad": undefined,
              "Program": undefined,
              "Type": "UnloadToCompletedMaterial",
              "UnloadIntoQueue": undefined,
            },
            "JobUnique": "xxx-offline-2018-01-01",
            "LastCompletedMachiningRouteStopIndex": undefined,
            "Location": Object {
              "CurrentQueue": undefined,
              "Face": 1,
              "Pallet": "5",
              "QueuePosition": undefined,
              "Type": "OnPallet",
            },
            "MaterialID": 102097,
            "PartName": "xxx",
            "Path": 1,
            "Process": 1,
            "Serial": "MBZG4",
            "SignaledInspections": Array [],
            "WorkorderId": undefined,
          },
        ],
        "pallet": Object {
          "CurrentPalletLocation": Object {
            "group": "L/U",
            "loc": "LoadUnload",
            "num": 2,
          },
          "FixtureOnPallet": "fix5",
          "NewFixture": undefined,
          "NumFaces": 1,
          "OnHold": false,
          "Pallet": "5",
          "PercentMoveCompleted": undefined,
          "TargetLocation": undefined,
        },
      },
    },
  },
}
`;
